pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java

```diff
@@ -0,0 +1,479 @@
+
+  private TableConfig _tableConfig;
+  protected FileSystem _fileSystem;
+
+  public SegmentPreprocessingJob(final Properties properties) {
+    _properties = properties;
+    _configuration = new Configuration();
+    setConf(_configuration);
+    Utils.logVersions();
+    logProperties();
+
+    if (_properties.getProperty(ENABLE_PARTITIONING) != null) {
+      _enablePartitioning = Boolean.parseBoolean(_properties.getProperty(ENABLE_PARTITIONING));
+    }
+
+    if (_properties.getProperty(ENABLE_SORTING) != null) {
+      _enableSorting = Boolean.parseBoolean(_properties.getProperty(ENABLE_SORTING));
+    }
+
+    if (!_enablePartitioning && _properties.getProperty(ENABLE_RESIZING) != null) {
+      _enableResizing = Boolean.parseBoolean(_properties.getProperty(ENABLE_RESIZING));
+    }
+
+    // get input/output paths.
+    _inputSegmentDir = properties.getProperty(PATH_TO_INPUT);
+    _preprocessedOutputDir = properties.getProperty(PATH_TO_OUTPUT);
+    _rawTableName = Preconditions.checkNotNull(_properties.getProperty(JobConfigConstants.SEGMENT_TABLE_NAME));
+
+    _pathToDependencyJar = _properties.getProperty(PATH_TO_DEPS_JAR, null);
+    _defaultPermissionsMask = _properties.getProperty(JobConfigConstants.DEFAULT_PERMISSIONS_MASK, null);
+
+    // Optional push location and table parameters. If set, will use the table config and schema from the push hosts.
+    String pushHostsString = _properties.getProperty(JobConfigConstants.PUSH_TO_HOSTS);
+    String pushPortString = _properties.getProperty(JobConfigConstants.PUSH_TO_PORT);
+    if (pushHostsString != null && pushPortString != null) {
+      _pushLocations =
+          PushLocation.getPushLocations(StringUtils.split(pushHostsString, ','), Integer.parseInt(pushPortString));
+    } else {
+      _pushLocations = null;
+    }
+
+    _logger.info("*********************************************************************");
+    _logger.info("enable.partitioning: {}", _enablePartitioning);
+    _logger.info("enable.sorting: {}", _enableSorting);
+    _logger.info("enable.resizing: {}", _enableResizing);
+    _logger.info("path.to.input: {}", _inputSegmentDir);
+    _logger.info("path.to.output: {}", _preprocessedOutputDir);
+    _logger.info("path.to.deps.jar: {}", _pathToDependencyJar);
+    _logger.info("push.locations: {}", _pushLocations);
+    _logger.info("*********************************************************************");
+  }
+
+  public void run() throws Exception {
+    if (!_enablePartitioning && !_enableSorting && !_enableResizing) {
+      _logger.info("Pre-computing job is disabled.");
+      return;
+    } else {
+      _logger.info("Starting {}", getClass().getSimpleName());
+    }
+
+    _fileSystem = FileSystem.get(_configuration);
+    final List<String> inputDataPath = Arrays.asList(_inputSegmentDir.split("\\s*,\\s*"));
+    Path outputDir = new Path(_preprocessedOutputDir);
+
+    if (_fileSystem.exists(outputDir)) {
```

Comment:
want to confirm, which output dir is this? is this the same output dir as the pbnj segment creation step? (assuming we set the segment creation step properly?)
or is it just one level in?
perhaps we can use a different default path to output, not sure what our default is - in segment creation, we actually don't require one, we have a default, so we should have a default here too, perhaps /preprocess?

Reply:
Good point. If output path isn't specified here, we should use a default path. I'll adjust the logic.

Reply:
Just checked the code. Actually in segment creation job, the output path should not be null. The path modification should be done when initializing the job, not inside the job.